(GH-15) Documentation added. (#53)
bbtsoftware/BBT.Maybe: docs/input/docs/usage/examples.md

---
Order: 20
Title: Examples

> Description: Examples of usages of BBT.Maybe

Reviewer: Example usage of BBT.Maybe

Author: Adapted.

---

## Usage of factory methods

> For the creation of maybe values factory methods are provided. Methods with arguments for reference are available (Some method). For the none case parameterless methods can be used (None method).

Reviewer: ```suggestion
Maybe values are created using factory methods provided by Maybe. Methods with arguments for reference are available (`Maybe.Some<T>`). For the none case parameterless methods can be used (`Maybe.None<T>()`).
```

Author: Adapted.

For nullable value types analogous methods postfixed with 'Struct' are provided.

```csharp

>     // Creation of a maybe for a declared reference type which could be instantiated or not.

Reviewer: Remove indentation as this indentates the code block's content.

Author: Done.

    var foo = new Foo();
    var maybeFoo = Maybe.Some(foo);

    // Creation of a maybe for a reference type which is null.
    var maybeFoo2 = Maybe.None<Foo>();

    // Creation of a maybe for a declared nullable value type which could be instantiated or not.
    int? number = 5;
    var maybeNumber = Maybe.SomeStruct<int>(number);

    // Creation of a maybe for a nullable value type which is null.
    var maybeNumber2 = Maybe.NoneStruct<int>();
```

## Optional properties

> Maybe is quite useful for optional properties. The default implementation of maybe represents the null case. Therefore optional properties of type maybe are representing the null case either and must not be initialized.

Reviewer: _too_ instead of _either_

Author: Adapted.

> Maybe is quite useful for optional properties. The default implementation of maybe represents the null case. Therefore optional properties of type maybe are representing the null case either and must not be initialized.

Reviewer: > The default implementation of maybe represents the null case

This sentence occurs multiple times. Can be placed in the introductionary paragraph of this document.

Author: Removed redundant sentence.

> Maybe is quite useful for optional properties. The default implementation of maybe represents the null case. Therefore optional properties of type maybe are representing the null case either and must not be initialized.

Reviewer: ```suggestion
Maybe is useful for optional properties. The default implementation of maybe represents the null case, therefore optional properties of `Maybe<T>` are representing the null case too and must not be initialized.
```

Author: Adapted.

```csharp
    /// Data class with mandatory property Foo and optional property Bar. Default of Bar is null.

>     public class Data()

Reviewer: Remove indentation as this indentates the code block's content.

Author: Done.

    {
        public Data(Foo foo)
        {
            Foo = foo;
        }

        public Foo { get; }
        public Maybe<Bar> MaybeBar { get; set; }
    }
```

## Optional method arguments

Maybe can be used as optional method arguments. The default implementation of maybe represents the null case.
Therefore optional method arguments can be declared using the default keyword.

```csharp

>     /// Data class with mandatory property A and optional property B. Default of B is null.

Reviewer: Remove indentation as this indentates the code block's content.

Author: Done.

    public void Foo(bool flag, Maybe<A> maybeA = default)
    {
        ...
    }
```

## Projection method

Maybe provides a projection method (Some method) with an argument of type Func<T, TProjected>.
The usage is similar to the null propagation operator.

```csharp

>     /// Class foo contains optional property bar.

Reviewer: Remove indentation as this indentates the code block's content.

Author: Done.

    public class Foo()
    {
        public Bar { get; set; } = null;
    }
```

 Usage of the maybe projection functionality:

```csharp

>     Maybe.Some(foo).Some(x => x.Bar).Do(x => ...);

Reviewer: Remove indentation as this indentates the code block's content.

Author: Done.
